onecodex/viz/_custom_plots/__init__.py

```diff
@@ -0,0 +1,91 @@
+from __future__ import annotations
+
+from typing import TYPE_CHECKING, Any
+import json
+
+from onecodex.utils import get_requests_session
+from onecodex.viz import configure_onecodex_theme
+from onecodex.exceptions import OneCodexException
+from .collection import SampleCollection, Samples
+from .models import PlotParams
+from .enums import SuggestionType
+
+if TYPE_CHECKING:
+    from pyodide.ffi import JsProxy
+
+
+def init():
+    configure_onecodex_theme()
+
+
+def plot(params: JsProxy, csrf_token: str) -> dict:
+    import js  # available from pyodide
+
+    params = PlotParams.model_validate(_convert_jsnull_to_none(params.to_py()))
+
+    # Cache the *unfiltered* SampleCollection, keyed by (<tag_or_project_uuid>, <metric>).
+    # It looks gross but this is how you cache in Pyodide ¯\_(ツ)_/¯
+    cache = globals().get("CUSTOM_PLOTS_CACHE", {})
+
+    uuid = None
+    type_ = None
+    if params.tag:
+        uuid = params.tag
+        type_ = SuggestionType.Tag
+    elif params.project:
+        uuid = params.project
+        type_ = SuggestionType.Project
+    else:
+        raise OneCodexException("Neither a tag nor project UUID was provided.")
+
+    key = (uuid, params.metric)
+    if key in cache:
+        collection = cache[key]
+    else:
+        base_url = js.self.location.origin
+        url = f"{base_url}/api/v2/custom-plots/sample-data"
+
+        samples = []
+        page = 1
+        while True:
```

Comment:
I would make it a pseudo infinitive loop (`for _ in range(10_000):`).

----

Feel free to ignore.